neuralmonkey/training_profiler.py

```diff
@@ -0,0 +1,81 @@
+# pylint: disable=unused-import
+from typing import List, Optional
+# pylint: enable=unused-import
+import time
+
+from neuralmonkey.logging import log, notice
+
+
+class TrainingProfiler:
```

Comment:
Je profiler dobrý jméno? _Profiler_ mi přijde, že jenom zaznamenává, jak dlouho, co trvá. Tohle ale rozhoduje o tom, co se kdy pustí. Takže podle mě je to mnohem víc _Timer_ nebo _Scheduler_.

Reply:
tadyta třída o ničem nerozhoduje. má nějaký floatový properties a zbytek metod vrací `None`. plus muj záměr s touhle třídou je rozšířit jí tak, aby dělala to, co dělá profiler.

Reply:
jediná divná věc, kterou profiler nedělá a kterou tahle třída dělá, je to logování na konci validace. což ale svým způsobem profilery taky dělaj..

Reply:
ty rozhodovátka jsou funkce, který tvoří ten konfig normalizér, ne tohle.

Reply:
I tak mi to nepřijde jako profiler, kterej je něco, co pouštíš, aby jsi něco debugoval. Z profileru zjistíš nějakej profile, na ten se podiváš a něco si myslíš. Tohle říká, jestli je čas trénovat nebo čas logovat.

Reply:
neřiká. řikat, jestli je čas trénovat nebo logovat by znamenalo, že tu bude nějaká funkce, která vrací boolean. ta tu neni. jen settery a nic víc.

"profil" jsou v tomhle případě dva seznamy mezivalidačních a validačních časů, který tahle třída sbírá (a zatim s nima nic nedělá).

Reply:
training loop se sem kouká, když chce zjistit, jestli se má logovat nebo validovat. ale jen si odtud bere čas posledního logu/validace a pak si volá svoje funkce, který to rozhodnou.

Reply:
Co se dá dělat s těma profilama dál, to je na jinej pull request. Může se to ukládat do souboru, mohlo by se to strkat do tensorboardu, může se to využít na nějaký online rady (jakože se třeba zjistí, že se časy validace natolik prodloužily, že už to trvá 3x dýl než samotný trénování.

Reply:
Asi nemá cenu to řešit donekonečna. Pořád mi profiler zní jako něco, co si jako člověk naschál pustim, abych zjistil, jak co dlouho trvá a ne jako něco, co se volá automaticky, aby zjistilo, kolik času uběhlo od nějaký události. Když uvidim tuhletu třídu, tak si řeknu: "Aha, to je něco, co si pustim, až nebudu vědět, co žere kolik času a normálně to používat nebudu." Ale jestli myslíš, že nikdo jinej si to myslet nebude, tak to tak asi může zůstat.